### PR TITLE
Backend: Add SigV4 middleware when creating client

### PR DESCRIPTION
**Pull request by fridgepoet**
**What this PR does / why we need it**:
This adds the necessary elements to sign AWS API requests to the backend http client in case the user has enabled SigV4 auth.

It was mistakenly removed in a recent PR https://github.com/grafana/opensearch-datasource/pull/118/files#diff-a34641cdaf9cc8aa70b90fd5a73157563f3f495d200c36387a6daba1036635d0L66. 

**Which issue(s) this PR fixes**:

May be related to https://github.com/grafana/opensearch-datasource/issues/167

**Testing notes**
The way we have verified this is by running a query with a data source which has enabled SigV4 auth (see plugin-provisioning repo). 
With the current code on `main`, we would get No Data with a backend flow query, for example by adding a server-side expression like type Math, reference to the first query, e.g. `$A`. 
With the changes from this PR, we are able to see the same data come back on a SSE query as described above.

### CODE REVIEW

File: pkg/opensearch/client/client_test.go [hunk 1]
```diff
@@ -437,3 +438,74 @@ func generateCertificate(t *testing.T, commonName string, ca *x509.Certificate,
...
+		}`),
+		DecryptedSecureJSONData: map[string]string{"sigV4AccessKey": "some_access_key", "sigV4SecretKey": "some_secret_key"},
+	})
+	require.NoError(t, err)
+
+	_, err = client.Get(server.URL)
+	assert.NoError(t, err)
+}
+
+func Test_newDatasourceHttpClient_sets_aoss_as_service_name_for_serverless(t *testing.T) {
```
Comment (fridgepoet) [inline pkg/opensearch/client/client_test.go:484]: This test is unrelated to this PR. It tests the "serverless" option in the JSON settings.

File: pkg/opensearch/client/client.go [hunk 2]
```diff
@@ -55,6 +60,28 @@ var newDatasourceHttpClient = func(ds *backend.DataSourceInstanceSettings) (*htt
 	return httpClient, nil
 }
 
+func sigV4Middleware() httpclient.Middleware {
+	return httpclient.NamedMiddlewareFunc("sigv4", func(opts httpclient.Options, next http.RoundTripper) http.RoundTripper {
+		rt, err := sigv4.New(&sigv4.Config{
+			Service:       opts.SigV4.Service,
+			AccessKey:     opts.SigV4.AccessKey,
```
[the remark this answers is not in the document]
Comment (sarahzinger) [reply pkg/opensearch/client/client.go:67]: ok figured it out I think that this sigv4.New is coming from https://github.com/grafana/grafana-aws-sdk/blob/main/pkg/sigv4/sigv4.go and that we only use the config details we need depending on the auth type: https://github.com/grafana/grafana-aws-sdk/blob/main/pkg/sigv4/sigv4.go#L243

makes sense please ignore me!

**Review by sarahzinger — APPROVED** [review]
🎉

File: pkg/opensearch/client/client_test.go (hunk 1 above)
Comment (sarahzinger) [reply pkg/opensearch/client/client_test.go:484]: imo if this comment helps explain the code to your pr reviewer now, it is the kind of thing that is helpful as a permanent incode comment, but I am a bit obsessive about putting comments everywhere that tend to grow stale so feel free to ignore me on this

File: pkg/opensearch/client/client.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (fridgepoet) [reply pkg/opensearch/client/client.go:67]: This is eventually called in the chain using the httpClientOptions created above (called with grafana-plugin-sdk-go's http client provider New right below). 

I may be wrong, but I think this is how this works: 
So `httpClientOptions, err := ds.HTTPClientOptions()` the config options are parsed from the JSON/data source config editor.
Then we add some customizations related to SigV4 right below it including this function you're asking about right here. It just takes something from "opts." But where is "opts" coming from, you ask? 

So when we call `httpClient, err := httpClientProvider.New(httpClientOptions)` here, in `New` we have `clientOpts := p.createClientOptions(opts...)` https://github.com/grafana/grafana-plugin-sdk-go/blob/v0.161.0/backend/httpclient/provider.go#L75
It seems to simply modify a little bit the httpClientOptions we passed in.
Because have passed in httpClientOptions in opensearch-datasource, we enter this case https://github.com/grafana/grafana-plugin-sdk-go/blob/v0.161.0/backend/httpclient/provider.go#L162 and
there we see that it combines the httpClientProvider middleware (anything from the sdk I suppose) with the SigV4 middleware which was already in `clientOpts.Middlewares`. 

Then it takes those mostly-same httpClientOptions and enters `New` here https://github.com/grafana/grafana-plugin-sdk-go/blob/v0.161.0/backend/httpclient/provider.go#L82

Follow `New` to `GetTransport` and you eventually get to the bit where we use the middleware we pass in. https://github.com/grafana/grafana-plugin-sdk-go/blob/v0.161.0/backend/httpclient/http_client.go#L94 Because the mostly-same httpClientOptions were passed in, that's where the `opts` you're asking about comes from.

Sorry this is not a very nice explanation -- I find this very hard to read and follow. Might be able to run this with a debugger to verify this flow.

[the remark this answers is not in the document]
Comment (fridgepoet) [reply pkg/opensearch/client/client.go:67]: This is eventually called in the chain using the httpClientOptions created above (called with grafana-plugin-sdk-go's http client provider New right below) and httpClientOptions comes from the config. For the env part, some of the code which is reading from the env is from grafana-aws-sdk ( `sigv4.New` call). 

I may be wrong, but I think this is how this works: 
So `httpClientOptions, err := ds.HTTPClientOptions()` the config options are parsed from the JSON/data source config editor.
Then we add some customizations related to SigV4 right below it including this function you're asking about right here. It just takes something from "opts." But where is "opts" coming from, you ask? 

So when we call `httpClient, err := httpClientProvider.New(httpClientOptions)` here, in `New` we have `clientOpts := p.createClientOptions(opts...)` https://github.com/grafana/grafana-plugin-sdk-go/blob/v0.161.0/backend/httpclient/provider.go#L75
It seems to simply modify a little bit the httpClientOptions we passed in.
Because have passed in httpClientOptions in opensearch-datasource, we enter this case https://github.com/grafana/grafana-plugin-sdk-go/blob/v0.161.0/backend/httpclient/provider.go#L162 and
there we see that it combines the httpClientProvider middleware (anything from the sdk I suppose) with the SigV4 middleware which was already in `clientOpts.Middlewares`. 

Then it takes those mostly-same httpClientOptions and enters `New` here https://github.com/grafana/grafana-plugin-sdk-go/blob/v0.161.0/backend/httpclient/provider.go#L82

Follow `New` to `GetTransport` and you eventually get to the bit where we use the middleware we pass in. https://github.com/grafana/grafana-plugin-sdk-go/blob/v0.161.0/backend/httpclient/http_client.go#L94 Because the mostly-same httpClientOptions were passed in, that's where the `opts` you're asking about comes from.

Sorry this is not a very nice explanation -- I find this very hard to read and follow. Might be able to run this with a debugger to verify this flow.

[the remark this answers is not in the document]
Comment (fridgepoet) [reply pkg/opensearch/client/client.go:67]: I realize I may not have even looked at the right part of your question, let me know if I misunderstood.
